pkg/services/alerting/eval_context.go

```diff
@@ -178,3 +181,55 @@ func getNewStateInternal(c *EvalContext) models.AlertStateType {
 
 	return models.AlertStateOK
 }
+
+// evaluateNotificationTemplateFields will treat the alert evaluation rule's name and message fields as Go text
+// templates, and execute the templates using data from the alert evaluation's tags
+func (c *EvalContext) evaluateNotificationTemplateFields() error {
+	if len(c.EvalMatches) < 1 {
+		return nil
+	}
+
+	msgTmpl, err := template.New("").Parse(c.Rule.Message)
+	if err != nil {
+		return err
+	}
+
+	templateDataMap := buildTemplateDataMap(c.EvalMatches)
+
+	var msgBuf bytes.Buffer
+	if err := msgTmpl.Execute(&msgBuf, templateDataMap); err != nil {
+		return err
+	}
+	c.Rule.Message = msgBuf.String()
+
+	titleTmpl, err := template.New("").Parse(c.Rule.Name)
+	if err != nil {
+		return err
+	}
+
+	var titleBuf bytes.Buffer
+	if err := titleTmpl.Execute(&titleBuf, templateDataMap); err != nil {
+		return err
+	}
+	c.Rule.Name = titleBuf.String()
+	return nil
+}
+
+// buildTemplateDataMap builds a map of alert evaluation tag names to a set of associated values (comma separated)
+func buildTemplateDataMap(evalMatches []*EvalMatch) map[string]string {
+	var result = map[string]string{}
+	for _, match := range evalMatches {
+		for tagName, tagValue := range match.Tags {
+			// skip duplicate values
+			if strings.Contains(result[tagName], tagValue) {
```

Comment:
AFAICT, using `strings.Contains` here is buggy, because it will match a substring of an existing value. F.ex., if you have "0.9990" in your string, "0.999" will be determined as a duplicate.